Fix Closure compilation

TimvdLippe opened the pull request:
We recently switched to use your polyfill for Promises instead of different Promise polyfill. As we are using Google Closure compiler, I noticed that this change introduced some warnings.

This PR adds some small necessary type annotations for closure to understand the source code and type-check appropriately. The fixes were:

1. Add `@constructor` to the constructors
2. Add type annotations for the state values defined in the Promise
3. Declare the `global` variable such that Closure understand it is declared, but still can use the `global` of Node

taylorhakes commented:
Hi @TimvdLippe, I really appreciate contributions, but please open an issue before creating a PR. This is a major change to the library and we need to discuss in depth to make sure the change makes sesne.

How do you work with other 3rd party libraries that don't have Closure compiler definitions? Are you able to create stubs like Typescript and Flowtype?  I generally don't like adding additional  inline comments/code for a specific type checker.

taylorhakes commented on src/polyfill.js:
```diff
@@ -1,6 +1,7 @@
 import Promise from './index';
 import promiseFinally from './finally';
 
+var global;
```
This shadows the global in Node JS and will cause it to not work correctly

TimvdLippe replied:
Suppressed the warning instead. Fixed :+1:

TimvdLippe commented:
@taylorhakes Normally I would open an issue before contributing, but I did not consider this to be a major change and thus thought opening a PR was easier. Sorry for that :cry: 

In general, JSDoc is the standard for specifying types in comments in JavaScript files. The Closure Compiler uses JSDoc types: http://usejsdoc.org/tags-type.html. TypeScript supports JSDoc as well: https://github.com/Microsoft/TypeScript/wiki/JSDoc-support-in-JavaScript (I added a check for this and fixed up some of the JSDoc so that `tsc` will work fine now as well) Sadly Flow does not support JSDoc just yet: https://github.com/facebook/flow/issues/5670

taylorhakes commented:
I categorize major changes as anything that isn't just a typo change or a small bug. I am ok with this change, but it looks like the build is failing. Would you mind fixing the build?

TimvdLippe commented:
@taylorhakes I was unable to figure out why the build was failing. I was using the exact same setup as Travis and it was passing locally... Not sure why TypeScript chokes on that one file on CI, but not locally. Anyhow, all green now. Thanks for your patience and review!

taylorhakes commented:
@TimvdLippe Sorry for the delay. I am ready to merge this PR, but it looks like you added a `package-lock.json`. Can you remove that file? I don't want the dependencies of this library to be locked.

TimvdLippe commented:
@taylorhakes Done 😄

git-jiby-me commented:
@taylorhakes is it possible to release a version already with these changes, as we use closure compiler as well, and will need this to use the library
